lib/rules/no-constant-condition.js

```diff
@@ -32,11 +32,15 @@ module.exports = function(context) {
                 return true;
 
             case "UnaryExpression":
-                return isConstant(node.argument);
+                return node.operator === "typeof" || isConstant(node.argument);
 
             case "BinaryExpression":
             case "LogicalExpression":
-                return isConstant(node.left) && isConstant(node.right) && node.operator !== "in";
+                return node.left.operator !== "typeof" &&
```

Comment:
Yes, I would use a second parameter (`inBooleanPosition`?) that is initially `true` but can be made `false` (and `true` again!) for certain sub-expressions. These should all be reported as constant conditions:

```js
if (typeof x);
if (void x);
if (typeof "constant" === "constant");
if (void x === void x);
if (0, typeof x);
if (!typeof x);
if (!!typeof x);
```